### ISSUE TITLE
TimelineAnimationEngine leaks memory (entries in _animations)

### ISSUE THREAD

**Issue by fhackenberger** [angular/angular#20585]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
When creating animations like this:
```
const myAnimation = this.animBuilder.build([
	style({ opacity: 1 }),
	animate('300 ms ease', style({ opacity: 0 }))
]);
const player = myAnimation.create(container.nativeElement);
player.onDone(() => { ...; player.destroy(); });
player.play();
```

This individual [AnimationMetadata](https://github.com/angular/angular/blob/7cf5e95ac9f0f2648beebf0d5bd9056b79946970/tools/public_api_guard/animations/animations.d.ts#L20) (i.e. the steps it needs to go through) is registered with [TimelineAnimationEngine](https://github.com/angular/angular/blob/master/packages/animations/browser/src/render/timeline_animation_engine.ts) stored in TimelineAnimationEngine._animations. However there is no code in that class that will ever clear entries from the map _animations. IDs for new animations are created within BrowserAnimationBuilder, by simply incrementing from the previous issues ID.

The callflow is like this:

BrowserAnimationBuilder.build() -> issueAnimationCommand() -> AnimationRenderer.setProperty(name: "@@XXXX:register") -> AnimationEngine.process() -> TimelineAnimationEngine.command() -> TimelineAnimationEngine.register()

## Expected behavior
The animation that is returned from build() needs to provide a destroy() method that it de-registers the animation itself.

## Minimal reproduction of the problem with instructions

[See this plnkr](http://embed.plnkr.co/rPW5xGXj3XUcq82WaoBQ/), open the preview in a separate (private) window in Chrome and let it run for a few animation cycles. Then take three memory heap snapshots with about 10 seconds in between. Select the third and select 'Objects allocated between Snapshot 1 and 2'. Observe that there is one retained _Object_ entry for each animation that was created using build().

## What is the motivation / use case for changing the behavior?
I stumbled upon this error, because from the API of AnimationBuilder, it was completely unclear that the animation would be centrally registered and cause memory to remain there during the whole life of the application.

The use case for creating lots of animations, is that for example the duration may vary according to user input, so there _must_ be a way to dispose of animations entirely for that use-case to be viable.

## Environment

<pre><code>
Angular version: 5.0.2

Browser:
- [x] Chrome (desktop) version 62
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [x] Firefox version 57
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: v6.10.1  
- Platform: Linux (Ubuntu 16.04.3) 

Others:

</code></pre>

**Comment by fhackenberger** [comment]
![screenshot from 2017-11-22 14-40-18](https://user-images.githubusercontent.com/910936/33130492-d59bf84e-cf93-11e7-9e27-c3a14d971795.png)

**Comment by woppa684** [comment]
I have the same issue with the TransitionAnimationEngine (in combination with routing). When I come back to the route that has animations they are added again and again to this statesByElement map and leaves a huges number of elements in the detached DOM.

![leak](https://user-images.githubusercontent.com/10152542/33165857-9a619e82-d038-11e7-86d7-db8cf808998d.png)

**Comment by alexhobbs** [comment]
Do you need anything to help reproduce this? We see this in a simple animating “cards” application where after one day of transitioning between cards every ten seconds the app crashes due to memory usage. Debugging this shows it to be the growing number of detached nodes. The workaround is to use native CSS where possible.

**Comment by DougJordan9** [comment]
Is there a solution to this problem? I think this is a major issue that essentially renders the animation library useless...

**Comment by DieterGehring** [comment]
I face the same issue with code below.
Is there any solution?
```
            // first build the animation
            let myAnimation = this._builder.build([
                style({ bottom: curPos }),
                animate("750ms ease-in", style({ bottom: this._lastPosition }))
            ]);

            // then create a player from it
            if (this.NeedleElement != null && this.NeedleElement.nativeElement) {
                let player = myAnimation.create(this.NeedleElement.nativeElement as HTMLElement);
                player.play();
            }

```

**Comment by ARadauer** [comment]
There is a very nice example for animations at the bottom of https://angular.io/guide/animations -> http://animationsftw.in/

This are the DOM Nodes after a while:
![image](https://user-images.githubusercontent.com/296251/96735830-41951200-13bc-11eb-83cf-53cd15b8b57a.png)

Don't know if its the same issue but it looks dodgy

**Comment by JeanMeche** [comment]
I've been investigating the memory leaks of the animation module. 

The issue reported by @woppa684  is the same as #24197 & #50533 : The cache of the `TransitionAnimationEngine` isn't freed in some situations. I've proposed a fix in #50929

The `TimelineAnimationEngine` is another one, lets track on this one here !

**Comment by BraianS** [comment]
Goodnight.
This issue can be closed because #24197 and #50533 are the same and are also close.

Thanks.
